File: clients/html5websocket/js/signal.apis.js
var signalAPIsHost = "http://localhost:8084/Signal/";

function sendSignal(action, token, fbref, fbdata){
    var targetUrl = signalAPIsHost + "otp-apis?token=" + token + "&fbref=" + fbref + "&uxt=" + new Date().getTime() + "&callback=?";
    $.ajax({
        url: targetUrl,
        type: "GET",
        dataType: "jsonp",
        contentType: "application/json;charset=utf-8",
        crossDomain: true
    }).done(function(data) {
        var status = data.status;
        var message = data.message;
        var code = data.code;
        if(status == 200){
            if(action == "signal.ack"){
                sendFeedbackSignal(code, token, fbref, fbdata); // *fbdata => *receiver
            }else if(action == "signal.var"){
                sendFeedbackVar(code, token, fbref, fbdata);
            }
        }
    });
}

function sendFeedbackSignal(code, token, fbref, receiver){
    var targetUrl = signalAPIsHost + "ack-apis?receiver=" + receiver + "&code=" + code + "&token=" + token + "&fbref=" + fbref + "&platform=html5&callback=?";
    $.ajax({
        url: targetUrl,
        type: "GET",
        dataType: "jsonp",
        contentType: "application/json;charset=utf-8",
        crossDomain: true
    }).done(function(data) {
        var status = data.status;
        var message = data.message;
        if(status == 200){
            $("#logs").append("<p>ACK APIs Message: " + message + "</p>");
        }
    });
}

function sendFeedbackVar(code, token, fbref, fbdata){
    var targetUrl = signalAPIsHost + "send-apis?code=" + code + "&token=" + token + "&fbref=" + fbref + "&data=" + fbdata + "&callback=?";
    $.ajax({
        url: targetUrl,
        type: "GET",
        dataType: "jsonp",
        contentType: "application/json;charset=utf-8",
        crossDomain: true
    }).done(function(data) {
        var status = data.status;
        var message = data.message;
        if(status == 200){
            $("#logs").append("<p>VAR APIs Message: " + message + "</p>");
        }
    });
}

function getNFSignal(uuid, secret, pointer, cause, condition){
    var targetUrl = signalAPIsHost + "signage-apis?uuid=" + uuid + "&secret=" + secret + "&uxt=" + new Date().getTime() + "&callback=?";
    $.ajax({
        url: targetUrl,
        type: "GET",
        dataType: "jsonp",
        contentType: "application/json;charset=utf-8",
        crossDomain: true
    }).done(function(data) {
        var status = data.status;
        var message = data.message;
        var signature = data.signature;
        if(status == 200){
            sendNQLAndGetNotificationData(signature, pointer, cause, condition);
        }
    });
}

function sendNQLAndGetNotificationData(signature, pointer, cause, condition){
    var nql = "SELECT ALL FROM " + pointer + " WHERE " + cause + " IS " + condition + " ON TODAY DEVICE ANY ORDER DATE DESC LIMIT 0,50";

    nql = encodeURIComponent(nql);
    var targetUrl = signalAPIsHost + "nql-apis?signature=" + signature + "&nql=" + nql;
    $.ajax({
        url: targetUrl,
        type: "GET",
        dataType: "jsonp",
        contentType: "application/json;charset=utf-8",
        crossDomain: true
    }).done(function(data) {
        $("#logs").append("<p>" + JSON.stringify(data) + "</p>");
    });
}

File: clients/java/Jignal/src/edu/sit/signal/apis/consumer/SignalAPIs.java
package edu.sit.signal.apis.consumer;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.Charset;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

/**
 *
 * @author Suthat
 */
public class SignalAPIs {
    
    private String signalAPIsHost = "http://localhost:8084/Signal/";
    
    public String requestOTP(String token, String fbref){ 
        String url = signalAPIsHost + "otp-apis?token=" + token + "&fbref=" + fbref + "&uxt=" + (System.currentTimeMillis() / 1000L);
        try {
            InputStream is = new URL(url).openStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String jsonText = "";
            StringBuilder sb = new StringBuilder();
            int cp;
            while ((cp = rd.read()) != -1) {
                sb.append((char) cp);
            }
            jsonText = sb.toString();
            try{
                JSONParser parser = new JSONParser();
                Object obj = parser.parse(jsonText);
                JSONObject jsonObject = (JSONObject) obj;
                long status = (Long) jsonObject.get("status");
                if(status == 200L){
                   return jsonObject.get("code").toString();
                }else{
                    System.out.println("requestOTP error on request >> " + status + "-" + jsonObject.get("message"));
                }
            }catch(Exception e){
                System.out.println("requestOTP error on json parsing >> " + e.getMessage());
            }
            is.close();
        } catch (Exception e) {
            System.out.println("requestOTP error on connection openning >> " + e.getMessage());
        }
        return "";
    }
    
    public String sendFeedbackSignal(String code, String token, String fbref, String receiver){ 
        String url = signalAPIsHost + "ack-apis?receiver=" + receiver + "&code=" + code + "&token=" + token + "&fbref=" + fbref + "&platform=cli";
        try {
            InputStream is = new URL(url).openStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String jsonText = "";
            StringBuilder sb = new StringBuilder();
            int cp;
            while ((cp = rd.read()) != -1) {
                sb.append((char) cp);
            }
            jsonText = sb.toString();
            try{
                JSONParser parser = new JSONParser();
                Object obj = parser.parse(jsonText);
                JSONObject jsonObject = (JSONObject) obj;
                long status = (Long) jsonObject.get("status");
                if(status == 200L){
                    return jsonObject.get("message").toString();
                }else{
                    System.out.println("sendFeebackSignal error on request >> " + status + "-" + jsonObject.get("message"));
                }
            }catch(Exception e){
                System.out.println("sendFeebackSignal error on json parsing >> " + e.getMessage());
            }
            is.close();
        } catch (Exception e) {
            System.out.println("sendFeebackSignal error on connection openning >> " + e.getMessage());
        }
        return "";
    }
    
    public String sendFeedbackVar(String code, String token, String fbref, String data){ 
        String url = signalAPIsHost+ "send-apis?code=" + code + "&token=" + token + "&fbref=" + fbref + "&data=" + data;
        try {
            InputStream is = new URL(url).openStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String jsonText = "";
            StringBuilder sb = new StringBuilder();
            int cp;
            while ((cp = rd.read()) != -1) {
                sb.append((char) cp);
            }
            jsonText = sb.toString();
            try{
                JSONParser parser = new JSONParser();
                Object obj = parser.parse(jsonText);
                JSONObject jsonObject = (JSONObject) obj;
                long status = (Long) jsonObject.get("status");
                if(status == 200L){
                    return jsonObject.get("message").toString();
                }else{
                    System.out.println("sendFeebackVar error on request >> " + status + "-" + jsonObject.get("message"));
                }
            }catch(Exception e){
                System.out.println("sendFeebackVar error on json parsing >> " + e.getMessage());
            }
            is.close();
        } catch (Exception e) {
            System.out.println("sendFeebackVar error on connection openning >> " + e.getMessage());
        }
        return "";
    }
    
    public String requestSignage(String uuid, String secret){ 
        String url = signalAPIsHost + "signage-apis?uuid=" + uuid + "&secret=" + secret + "&uxt=" + (System.currentTimeMillis() / 1000L);
        try {
            InputStream is = new URL(url).openStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String jsonText = "";
            StringBuilder sb = new StringBuilder();
            int cp;
            while ((cp = rd.read()) != -1) {
                sb.append((char) cp);
            }
            jsonText = sb.toString();
            try{
                JSONParser parser = new JSONParser();
                Object obj = parser.parse(jsonText);
                JSONObject jsonObject = (JSONObject) obj;
                long status = (Long) jsonObject.get("status");
                if(status == 200L){
                   return jsonObject.get("signature").toString();
                }else{
                    System.out.println("requestSignage error on request >> " + status + "-" + jsonObject.get("message"));
                }
            }catch(Exception e){
                System.out.println("requestSignage error on json parsing >> " + e.getMessage());
            }
            is.close();
        } catch (Exception e) {
            System.out.println("requestSignage error on connection openning >> " + e.getMessage());
        }
        return "";
    }
    
    public JSONObject sendNQLAndGetNotificationData(String signature, String pointer, String cause, String condition){
        JSONObject jsonObject = null;  
        try{
            String nql = "SELECT ALL FROM " + pointer + " WHERE " + cause + " IS " + condition + " ON TODAY DEVICE ANY ORDER DATE DESC LIMIT 0,50";
            nql = URLEncoder.encode(nql, "UTF-8");
               
            String url = signalAPIsHost + "nql-apis?signature=" + signature + "&nql=" + nql;
            
            InputStream is = new URL(url).openStream();
            BufferedReader rd = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));
            String jsonText = "";
            StringBuilder sb = new StringBuilder();
            int cp;
            while ((cp = rd.read()) != -1) {
                sb.append((char) cp);
            }
            jsonText = sb.toString();
            System.out.println(jsonText);
            try{
                JSONParser parser = new JSONParser();
                Object obj = parser.parse(jsonText);
                jsonObject = (JSONObject) obj;
            }catch(Exception e){
                System.out.println("error on json parsing >> " + e.getMessage());
            }
        }catch(Exception e){
            System.out.println("errror on url openning >> " + e.getMessage());
        }
        return jsonObject;
    }
}
